File: backend/app/schemas/analytics.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AnalyticsBase(BaseModel):
    lead_id: int
    page_time_seconds: int = Field(0, ge=0)
    button_clicks: dict[str, Any] = Field(default_factory=dict)
    cursor_heatmap: dict[str, Any] = Field(default_factory=dict)
    return_visits: int = Field(0, ge=0)
    session_data: dict[str, Any] = Field(default_factory=dict)


class AnalyticsCreate(AnalyticsBase):
    pass


class AnalyticsUpdate(BaseModel):
    page_time_seconds: int | None = Field(None, ge=0)
    button_clicks: dict[str, Any] | None = None
    cursor_heatmap: dict[str, Any] | None = None
    return_visits: int | None = Field(None, ge=0)
    session_data: dict[str, Any] | None = None


class AnalyticsResponse(AnalyticsBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime

File: frontend/src/types/analytics.ts
export interface AnalyticsCreate {
  lead_id: number;
  page_time_seconds: number;
  button_clicks: Record<string, number>;
  cursor_heatmap: Record<string, number>;
  return_visits: number;
  session_data: SessionData;
}

export interface AnalyticsResponse extends AnalyticsCreate {
  id: number;
  created_at: string;
}

export interface AnalyticsUpdate {
  page_time_seconds?: number;
  button_clicks?: Record<string, number>;
  cursor_heatmap?: Record<string, number>;
  return_visits?: number;
  session_data?: Partial<SessionData>;
}

export interface SessionData {
  field_focus_times: Record<string, number>;
  field_focus_order: string[];
  form_start_time: number | null;
  form_submit_time: number | null;
  total_form_time_ms: number;
  page_visibility_changes: number;
  screen_width: number;
  screen_height: number;
  user_agent: string;
}
